Blog Post: Custom Resource Monitoring (#524)
strimzi/strimzi.github.io: _posts/2025-09-05-monitoring-of-custom-resources.md

---
layout: post
title:  "Monitoring of custom resources"
date: 2025-09-05
author: sebastian_gaiser_hetzner
---

Running Kafka on Kubernetes with Strimzi is straightforward.
But getting a complete, consistent view of everything in your Kubernetes and Kafka cluster(s) can still be hard.

> Kafka topics and users are often owned by different teams and configured differently, especially when sharing a Kafka cluster e.g. across a company.

Reviewer: ```suggestion
Kafka topics and users are often owned by different teams and configured differently, especially when a Kafka cluster is shared across a company (for example, by multiple teams).
```

As usage grows, gathering information and troubleshooting gets harder.

This post shows a practical, low‑friction way to close monitoring gaps using [kube-state-metrics](https://github.com/kubernetes/kube-state-metrics) so you can track every Strimzi custom resource (CR) with a simple, consistent, Kubernetes‑native approach.

### What Strimzi exposes today

> All Strimzi operators (Cluster, User, and Topic Operator) expose metrics in Prometheus compatible format.

Reviewer: ```suggestion
All Strimzi operators (Cluster, User, and Topic Operator) expose metrics in Prometheus-compatible format.
```

> However, today only the Cluster Operator exports a metric (`strimzi_resource_state`) that describes the state of custom resources (ready or not).

Reviewer: ```suggestion
However, only the Cluster Operator exposes a metric (`strimzi_resource_state`) that reports the state of custom resources (ready or not).
```

It provides this metric for:

- `Kafka`
- `KafkaConnect`
- `KafkaBridge`
- `KafkaMirrorMaker2`
- `KafkaRebalance`

It does not provide this metric for:

- `KafkaNodePool`
- `KafkaConnector`

> - `StrimziPodSet`

Reviewer 2: I am not sure I would mention `StrimziPodSet` here because it's an internal resource and I would not expect users to be able monitoring it. I would focus on the users facing custom resources.

Reviewer 3: @ppatierno Why do you expect that users would not be able to monitor it? I'm not sure if StrimziPodSets give users some useful information (it is tracking some rolling update progress, but not sure it is useful on its own). But I do not see why they would not be able to monitor it.

Keep in mind it is also what is being annotated for manual rolling updates etc. So while the users are not expected and should not create or edit them (other than the annotation), it is expected that they know it exists and can monitor it.

The User and Topic Operators also do not provide this metric for `KafkaUser` and `KafkaTopic`.

> In day‑to‑day work, misconfigurations are happening to often like a typo in a topic setting, or using a deprecated field that was removed the last release.

Reviewer: ```suggestion
During routine work, it's easy to introduce misconfigurations, from a simple typo in a topic setting to using a property that was removed in the last release.
```

These small issues can trigger long debugging sessions. Often the fix is simple, but finding it is not.

### The simple fix: kube-state-metrics for CRDs

In general, kube-state-metrics can read any Kubernetes object, including custom resources, and export info‑style metrics with labels you control.
This gives you:

- A complete inventory of all Strimzi CRs
- Consistent metric labels across teams
- A Kubernetes‑native approach that fits existing tooling (Prometheus)

In short, you increase monitoring coverage without changing Strimzi or writing custom exporters.

### How to deploy

> In gerneral, Strimzi provides plain Kubernetes manifests in the examples directory.

Reviewer: ```suggestion
In general, Strimzi provides plain Kubernetes manifests in the examples directory.
```

                message: "Strimzi KafkaAccess {{`{{ $labels.name }}`}} is not ready"
            - alert: KafkaAccessDeprecated
              expr: strimzi_kafka_access_resource_info{deprecated="Warning"}
              for: 15m
              labels:
                severity: warning
              annotations:
                message: "Strimzi KafkaAccess {{`{{ $labels.name }}`}} contains a deprecated configuration"
```

Then you can simply install kube-state-metrics by executing:

```bash
helm install strimzi-kube-state-metrics oci://ghcr.io/prometheus-community/charts/kube-state-metrics --version 6.3.0 -f my-strimzi-kube-state-metrics-values.yaml
```

### Verify it works

Once Prometheus is scraping kube-state-metrics, try these queries to confirm it’s working.

Get an overview about your `Kafka` clusters:

```
sum by (namespace, name, cluster_id, ready) (strimzi_kafka_resource_info)
```

Example output:

```
{cluster_id="123456789", name="my-first-kafka",  namespace="kafka-one",   ready="True"}
{cluster_id="abcdefghj", name="my-second-kafka", namespace="kafka-two",   ready="True"}
{cluster_id="zyxwvutsr", name="my-third-kafka",  namespace="kafka-three", ready="False"}
```

> Get an overview of `KafkaTopics` with partitions and replicas:

Reviewer 2: ```suggestion
Get an overview of `KafkaTopic`(s) with partitions and replicas:
```

```
sum by (name, partitions, replicas) (strimzi_kafka_topic_resource_info)
```

Example output:

```
{name="my-first-kafka-topic",  partitions="1",  replicas="1"}
{name="my-second-kafka-topic", partitions="12", replicas="3"}
{name="my-third-kafka-topic",  partitions="60", replicas="3"}
```

> Get an overview of `KafkaUsers` with username:

Reviewer 2: ```suggestion
Get an overview of `KafkaUser`(s) with username:
```

It does not provide this metric for:

- `KafkaNodePool`
- `KafkaConnector`
- `StrimziPodSet`

The User and Topic Operators also do not provide this metric for `KafkaUser` and `KafkaTopic`.

During routine work, it's easy to introduce misconfigurations, from a simple typo in a topic setting to using a property that was removed in the last release.
These small issues can trigger long debugging sessions. Often the fix is simple, but finding it is not.

### The simple fix: use kube-state-metrics for CRDs

Instead of relying on the limited `strimzi_resource_state` metric, you can use kube-state-metrics to cover _all_ Strimzi custom resources.

kube-state-metrics can read any Kubernetes object, including custom resources, and export Info-style metrics with labels you control. This approach gives you:

- A complete inventory of all Strimzi CRs
- Consistent metric labels across teams
- A Kubernetes‑native approach that fits existing tooling (Prometheus)

In short, you increase monitoring coverage without changing Strimzi or writing custom exporters.

### How to deploy

In general, Strimzi provides plain Kubernetes manifests in the examples directory.

#### Option A — Apply manifests

> In [Strimzi 0.48.0](https://github.com/strimzi/strimzi-kafka-operator/releases/tag/0.48.0), examples are included to deploy kube-state-metrics with the corresponding:
> 
> - RBAC resources (`ServiceAccount`, `RoleBinding`, `ClusterRoleBinding`)
> - A `ConfigMap` for CR configuration
> - The `Deployment`
> - A `ServiceMonitor` for [Prometheus Operator](https://prometheus-operator.dev/) to scrape itself
> 
> [Examples directory for kube-state-metrics](https://github.com/strimzi/strimzi-kafka-operator/tree/0.48.0/examples/metrics/kube-state-metrics).

Reviewer 4: I'm assuming the kube-state-metrics examples will be including in all Strimzi versions going forwards. If that is the case I would suggest the following tweaks to wording and links to make that clear:
```suggestion
From [Strimzi 0.48.0](https://github.com/strimzi/strimzi-kafka-operator/releases/tag/0.48.0) onwards, examples are included to deploy kube-state-metrics with the corresponding:

- RBAC resources (`ServiceAccount`, `RoleBinding`, `ClusterRoleBinding`)
- A `ConfigMap` for CR configuration
- The `Deployment`
- A `ServiceMonitor` for [Prometheus Operator](https://prometheus-operator.dev/) to scrape itself

[Examples directory for kube-state-metrics](https://github.com/strimzi/strimzi-kafka-operator/tree/main/examples/metrics/kube-state-metrics).
```
